account_payment_mode_auto_reconcile/models/account_move.py

```diff
@@ -0,0 +1,164 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from operator import itemgetter
+
+from odoo import _, api, fields, models
+
+
+class AccountMove(models.Model):
+
+    _inherit = "account.move"
+
+    # Allow changing payment mode in open state
+    # TODO: Check if must be done in account_payment_partner instead
+    payment_mode_id = fields.Many2one(
+        states={
+            "draft": [("readonly", False)],
+            "posted": [("readonly", False)]
+        })
+    payment_mode_warning = fields.Char(
+        compute="_compute_payment_mode_warning",
+    )
+    display_payment_mode_warning = fields.Boolean(
+        compute="_compute_payment_mode_warning",
+    )
+
+    @api.model_create_multi
+    def create(self, vals):
+        res = super(AccountMove, self).create(vals)
+        for invoice in res:
+            if invoice.move_type != "out_invoice":
+                continue
+            if not invoice.payment_mode_id.auto_reconcile_outstanding_credits:
+                continue
+            partial = invoice.payment_mode_id.auto_reconcile_allow_partial
+            invoice.with_context(
+                _payment_mode_auto_reconcile=True
+            ).auto_reconcile_credits(partial_allowed=partial)
+        return res
+
+    def write(self, vals):
+        res = super(AccountMove, self).write(vals)
+        if "payment_mode_id" in vals or "state" in vals:
```

Comment:
In Odoo16, only invoice with [posted status](https://github.com/odoo/odoo/blob/16.0/addons/account/models/account_move.py#L1010)would list potential payments can be reconciled, so we only auto reconciled after posting the invoice, draft invoice will remove existing reconciled payment